lib/rucio/daemons/common.py

```diff
@@ -23,22 +23,133 @@
+        partition_wait_time: int = 1,
+        daemon_name: str = "undefined_daemon",
+    ) -> None:
+        """
+        :param once: Only execute one iteration.
+        :param total_workers: Number of total workers.
+        :param sleep_time: Sleep time between daemon iterations.
+        :param partition_wait_time: Time to wait for database partition rebalancing before starting the actual daemon loop.
+        :param daemon_name: Name of daemon that is constructed.
+        """
+        self.once = once
+        self.total_workers = total_workers
+        self.sleep_time = sleep_time
+        self.partition_wait_time = partition_wait_time
+        self.daemon_name = daemon_name
+        self.graceful_stop = threading.Event()
+        setup_logging(process_name=daemon_name)
+
+    @staticmethod
+    def _pre_run_checks() -> None:
+        """
+        Checks to run before daemon execution
+        """
+        if rucio.db.sqla.util.is_old_db():
+            raise DatabaseException("Database was not updated, daemon won't start")
+
+    @abstractmethod
+    def _run_once(
+        self, heartbeat_handler: "HeartbeatHandler", **_kwargs
+    ) -> tuple[bool, Any]:
+        """
+        Daemon-specific logic (to be defined in child classes) for a single iteration
+
+        :param heartbeat_handler: Handler to set and manage the heartbeat for this execution.
+
+        :returns: Tuple of (must_sleep, ret_value).
+                  must_sleep: set to True to signal to the calling context that it must sleep before next execution, False otherwise
+                  ret_value: Daemon-specific return value
+        """
+        pass
+
+    def _call_daemon(self, activities: Optional[list[str]] = None) -> None:
+        """
+        Run the daemon loop and call the function _run_once at each iteration
+        """
+
+        run_once_fnc = functools.partial(self._run_once)
+
+        daemon = db_workqueue(
+            once=self.once,
+            graceful_stop=self.graceful_stop,
+            executable=self.daemon_name,
+            partition_wait_time=self.partition_wait_time,
+            sleep_time=self.sleep_time,
+            activities=activities,
+        )(run_once_fnc)
+
+        for _ in daemon():
+            pass
+
+    def run(self) -> None:
+        """
+        Run the daemon.
+        """
+        self._pre_run_checks()
+
+        logging.info(f"{self.daemon_name}: starting threads")
+        thread_list = [
+            threading.Thread(target=self._call_daemon)
+            for _ in range(0, self.total_workers)
+        ]
+        [t.start() for t in thread_list]
+
+        if not self.once:
+            logging.info(f"{self.daemon_name}: waiting for interrupts")
+
+        # Interruptible joins require a timeout.
+        while thread_list:
+            thread_list = [
+                thread.join(timeout=3.14)
+                for thread in thread_list
+                if thread and thread.is_alive()
+            ]
```

Comment:
This whole code block was copied from Reaper:

```
    # Interruptible joins require a timeout.
    while threads_list:
        threads_list = [thread.join(timeout=3.14) for thread in threads_list if thread and thread.is_alive()]
```

In general, every daemon involved in the refactoring had a loop of this kind, e.g. in Undertaker:

```
       # Interruptible joins require a timeout.
        while threads[0].is_alive():
            [t.join(timeout=3.14) for t in threads]
```

Based on the fact that `threading.Thread.join()` returns `None`, then I think the version of the loop present in Undertaker is the one that's actually correct, given that it loops as long as the first thread is alive, right?

Reply:
I’m afraid I don’t know. That’s why I’m genuinely asking, because you’re (rightfully) copying existing implementations, and maybe they copied it over from somewhere else, but I wonder if anyone really understands them.

So, since the long-term goal of your effort is to implement this exactly once, I am taking the opportunity to try to understand exactly what it is supposed to do.

Reply:
The more I look at the existing implementations, the more perplexed I am. Won’t the loop in the Reaper exit after one iteration?

```python
>>> import threading
>>> import time
>>> def worker():
...     while True:
...         time.sleep(60)
... 
>>> threads_list = [threading.Thread(target=worker) for _ in range(2)]
>>> for thread in threads_list:
...     thread.start()
... 
>>> while threads_list:
...     print(threads_list)
...     threads_list = [thread.join(timeout=0.1) for thread in threads_list if thread and thread.is_alive()]
... 
[<Thread(Thread-1 (worker), started 140304535520960)>, <Thread(Thread-2 (worker), started 140304525035200)>]
[None, None]
>>> # Huh?
```

Reply:
Following my own little tests, I am starting to question the need for a timeout and whether the try-except structure we have have in the daemons for keyboard interrupts is working how it is supposed to.

I am also looking for reproducible example as to why the following won’t do:

```python
for thread in threads:
    thread.join()
```

Reply:
I’m getting different results in Python 3.6 and 3.9. :shrug: I think this further drives the point that we implemented a solution in Python 2, based on the situation at the time, and kept it as it is without checking if it still makes sense.

Reply:
Thanks for looking into this! I'm not sure what's the best way forward here - the different results in different Python versions are a bit worrying (though AFAIK Rucio requires Python 3.9 to run, so at least from the user's perspective it will be consistent), but it might also be one of those things that "if it works now, better not touch it" (even though it doesn't really work).

It's definitely a carryover from Python 2 in some way, I found [this commit](https://github.com/rucio/rucio/commit/9fa89245fe468b3ce9b67cfa9dc06bac4c1b77e6) under [this issue](https://github.com/rucio/rucio/issues/1924) for Python 3 compatibility with daemons, and that logic was left untouched

Reply:
I would adjust this. We should take this as an opportunity to renew the code here and don't carry outdated (not to say wrong) choices further in the future.

Reply:
But the question remains: should we implement Ctrl+C leading to a graceful stop?